compiler/src/dotty/tools/dotc/transform/DropBreaks.scala

```diff
@@ -0,0 +1,213 @@
+ *  Drops `try` statements on breaks if no other uses of its label remain.
+ *  A Break throw with a `Label` created by some enclosing boundary is replaced
+ *  with a labeled return if
+ *
+ *    - the throw and the boundary are in the same method, and
+ *    - there is no try expression inside the boundary that encloses the throw.
+ */
+class DropBreaks extends MiniPhase:
+  import DropBreaks.*
+
+  import tpd._
+
+  override def phaseName: String = DropBreaks.name
+
+  override def description: String = DropBreaks.description
+
+  override def runsAfterGroupsOf: Set[String] = Set(ElimByName.name)
+    // we want by-name parameters to be converted to closures
+
+  private object LabelTry:
+
+    object GuardedThrow:
+
+      /** `(ex, local)` provided `expr` matches
+       *
+       *      if ex.label.eq(local) then ex.value else throw ex
+       */
+      def unapply(expr: Tree)(using Context): Option[(Symbol, Symbol)] = stripTyped(expr) match
+        case If(
+          Apply(Select(Select(ex: Ident, label), eq), (lbl @ Ident(local)) :: Nil),
+          Select(ex2: Ident, value),
+          Apply(throww, (ex3: Ident) :: Nil))
+        if label == nme.label && eq == nme.eq && local == nme.local && value == nme.value
+            && throww.symbol == defn.throwMethod
+            && ex.symbol == ex2.symbol && ex.symbol == ex3.symbol =>
+          Some((ex.symbol, lbl.symbol))
+        case _ =>
+          None
+    end GuardedThrow
+
+    /** `(local, body)` provided `tree` matches
+     *
+     *      try body
+     *      catch case ex: Break =>
+     *        if ex.label.eq(local) then ex.value else throw ex
+     */
+    def unapply(tree: Tree)(using Context): Option[(Symbol, Tree)] = stripTyped(tree) match
+      case Try(body, CaseDef(pat @ Bind(_, Typed(_, tpt)), EmptyTree, GuardedThrow(exc, local)) :: Nil, EmptyTree)
+      if tpt.tpe.isRef(defn.BreakClass) && exc == pat.symbol =>
+        Some((local, body))
+      case _ =>
+        None
+  end LabelTry
+
+  private object BreakBoundary:
+
+    /** `(local, body)` provided `tree` matches
+     *
+     *      { val local: Label[...] = ...; <LabelTry(local, body)> }
+     */
+    def unapply(tree: Tree)(using Context): Option[(Symbol, Tree)] = stripTyped(tree) match
+      case Block((vd @ ValDef(nme.local, _, _)) :: Nil, LabelTry(caughtAndRhs))
+      if vd.symbol.info.isRef(defn.LabelClass) && vd.symbol == caughtAndRhs._1 =>
+        Some(caughtAndRhs)
+      case _ =>
+        None
+  end BreakBoundary
+
+  private object BreakThrow:
+
+    /** `(local, arg)` provided `tree` matches inlined
+     *
+     *    val Label_this: ... = local
+     *    throw new Break[...](Label_this, arg)
+     */
+    def unapply(tree: Tree)(using Context): Option[(Symbol, Tree)] = tree match
+      case Inlined(_,
+        (vd @ ValDef(label_this1, _, id: Ident)):: Nil,
+        Apply(throww, Apply(constr, Inlined(_, _, Ident(label_this2)) :: arg :: Nil) :: Nil))
```

Comment:
`Inlined(_, _, Ident(label_this2))` is too specific. We should have a `Label` extractor that can deal with any nesting of `Inlined` and `Block`. See 1b6f1824ca8a83ce6b5564d04ab69027ad9abba8.

Reply:
We now treat breaks as non-inlined calls, so the problem goes away